test/apps/iodemo/io_demo.cc

```diff
@@ -650,20 +650,33 @@ class P2pDemoCommon : public UcxContext {
             }
         }
 
-        void init(size_t data_size, void *ext_buf, bool validate)
+        void init(size_t data_size, BufferMemoryPool<Buffer> &chunk_pool,
+                  void *ext_buf, bool validate)
         {
             assert(ext_buf != NULL);
             assert(_iov.empty());
             assert(_extra_buf == NULL);
+            assert(_bogus_buf == NULL);
 
             _data_size = data_size;
             _extra_buf = ext_buf;
+            _bogus_buf = new Buffer(NULL, 0, chunk_pool, UCS_MEMORY_TYPE_HOST, NULL,
```

Comment:
shall we pass `_ext_buf` as the first parameter to `Buffer` costructor?

Reply:
does it fit the "80 symbols per line" rule?

Reply:
> shall we pass `_ext_buf` as the first parameter to `Buffer` costructor?

no otherwise _extra_buf would be freed in Buffer deconstructor.

> does it fit the "80 symbols per line" rule?

done

Reply:
>>shall we pass _ext_buf as the first parameter to Buffer constructor?

>no otherwise _extra_buf would be freed in Buffer deconstructor.

but wasn't the idea to wrap `_extra_buf` by `Buffer` to return it from `operator[]`?

Reply:
_extra_buf was introduced because we found it was hard to be wrapped in Buffer when we were working on https://github.com/openucx/ucx/pull/7139. Buffer manages a buffer allocated/initialized/freed by io_demo while _extra_buf stores the pointer of a buffer managed by UCX. BufferIov has 2 versions of init() and member functions like validate(), release() need to consider both cases. the bogus Buffer idea was to support operator[] for _extra_buf because currently it caused segment fault.

Reply:
yes, I understand.
but how Buffer uses `_extra_buf`? or why `_extra_buf` needed now?

Reply:
`_extra_buf` is used when receiving active messages in eager mode where buffer passed in am receive callback could be used directly to avoid additional memcpy to user provided buffer. That is the purpose of https://github.com/openucx/ucx/pull/7139.
using Buffer means using user provided buffer while using `_extra_buf` means using UCX provided buffer. BufferIov need to consider both cases. operator[] didn't consider `_extra_buf` that's why we are here.

Reply:
thanks for the explanation!
now I see the idea.

but anyway it sounds like still problematic, we have _extra_buf and Buffer which are not correlate to each other.
wdyt about something like - https://github.com/dmitrygx/ucx/tree/topic/io_demo/am_eager_bogus_buffer? It removes `_extra_buf` as no longer needed and utilizes your idea to have bogus `Buffer`

Reply:
By supporting the zcopy case in `Buffer `and `BufferMemoryPool`, `BufferIov` looks more cleaner.

Reply:
>By supporting the zcopy case in Buffer and BufferMemoryPool, BufferIov looks more cleaner.

but `operator[]` returns `Buffer` which has buffer=NULL which is not expected, since passed `_extra_buf` isn't `NULL`

Reply:
I'd expect that all methods of `BufferIov` work in the same way for real IOV of `Buffer`s or bogus `Buffer`.
Until it is not true, the `BufferIov` is not _healthy_

Reply:
please checkout the latest commit

Reply:
thanks!